### ISSUE TITLE
calculate LPC coefficient from cepstrum in Python

### ISSUE THREAD

**Issue by tuanad121** [mozilla/LPCNet#93]
I have an issue when translating the C code in celt_lpc.c to Python. The Python LPC coefficients is very small compared to  C LPC coefficients. I attached my notebook with all steps here.
[cep2lpc.ipynb.zip](https://github.com/mozilla/LPCNet/files/3992072/cep2lpc.ipynb.zip).
Any helps would be very appreciated. 
Thanks for spending your time on the problem.

**Comment by auzxb** [comment]
Hi, have you solved the problem?

**Comment by auzxb** [comment]
I solved your problem by modifying this code
```
def _celt_lpc(ac, p):
    error = ac[0]
    lpc = np.zeros(p)
    rc = np.zeros(p).astype(float)
    if ac[0] != 0:
        for i in range(p):
            rr = 0
            for j in range(i):
                rr += lpc[j] * ac[i - j]
            rr += (ac[i + 1])
            r = -rr / error
            rc[i] = r
            lpc[i] = (r)
            for j in range(int((i+1) / 2)):
                tmp1 = lpc[j]
                tmp2 = lpc[i-1-j]
                lpc[j]     = tmp1 + (r * tmp2)
                lpc[i-1-j] = tmp2 + (r * tmp1)
            error -= ((r * r) * error)
            if error < (ac[0]  / (2**10)):
                break
            if (error < 0.001 * ac[0]):
                break
    return error, lpc, rc

```

**Comment by tuanad121** [comment]
@auzxb awesome. I think with the script we can modify the test_lpcnet.py to synthesize speech from cepstrum like what's in test_lpcnet.c. But I'm not sure if it's necessary. In my case, the results from test_lpcnet.c are noisier than those from test_lpcnet.py.

**Comment by auzxb** [comment]
Yes! Just for convenience, I'm also trying to convert the code to python version, and compare the  results.
In the original arch.h file of this project, the Macro maybe a bug.
`#define PSHR(a,shift)   (a)
#define SHR(a,shift)    (a)
#define SHL(a,shift)    (a)
#define SATURATE(x,a)   (x)
#define SATURATE16(x)   (x)
 `

**Comment by HanzhengLi** [comment]
> In the original arch.h file of this project, the Macro maybe a bug.
> 
> `#define PSHR(a,shift) (a) #define SHR(a,shift) (a) #define SHL(a,shift) (a) #define SATURATE(x,a) (x) #define SATURATE16(x) (x) `

If I have read the code correctly, it is not a bug. You need to have these functions properly defined only if the program is to be run on a fixed-point CPU.




> @auzxb awesome. I think with the script we can modify the test_lpcnet.py to synthesize speech from cepstrum like what's in test_lpcnet.c. But I'm not sure if it's necessary. In my case, the results from test_lpcnet.c are noisier than those from test_lpcnet.py.

Interesting, are there any confirmed cases you can share? Thanks!

**Comment by tuanad121** [comment]
@HanzhengLi: I incorporated the centrum to lpc into test_lpc.py and the synthesized speech is good. I haven't looked carefully so it makes me a little bit worried about it too good to be true :D :D . Here is the script if you want to take a look. 
[test_lpcnet2.py.zip](https://github.com/mozilla/LPCNet/files/4379069/test_lpcnet2.py.zip)

**Comment by HanzhengLi** [comment]
@tuanad121 Great work! Though I think synthesis in C should do almost exactly the same, except for approximation of nonlinear functions and the randomness in sampling. Still, good to have it work in python.

**Comment by JohnHerry** [comment]
> @auzxb awesome. I think with the script we can modify the test_lpcnet.py to synthesize speech from cepstrum like what's in test_lpcnet.c. But I'm not sure if it's necessary. In my case, the results from test_lpcnet.c are noisier than those from test_lpcnet.py.

We had got the opposite result. We are traing the front-end with Tacotron-2 with the project of  https://github.com/MlWoo/LPCNet.  The vocode from test_lpcnet.c is better then test_lpcnet.py; pcm from test_lpcnet.py is noisy!
